controllers/cluster_controller.go

```diff
@@ -199,6 +142,99 @@ func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *capi.Cluster
 	return requeue(), nil
 }
 
+func deleteVintageCluster(ctx context.Context, log logr.Logger, client ctrlclient.Client, cluster *capi.Cluster) error {
+	log.Info(fmt.Sprintf("Cluster %s/%s is being deleted", cluster.Namespace, cluster.Name))
+	if err := client.Delete(ctx, cluster, ctrlclient.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+		log.Error(err, fmt.Sprintf("unable to delete cluster %s/%s", cluster.Namespace, cluster.Name))
+		ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+		return err
+	}
+	log.Info(fmt.Sprintf("Cluster %s/%s was deleted", cluster.Namespace, cluster.Name))
+	SuccessTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+	return nil
+}
+
+func deleteClusterApp(ctx context.Context, log logr.Logger, client ctrlclient.Client, cluster *capi.Cluster) error {
+	// CAPI-based cluster but without Helm annotation? weird! should not happen; if do, we have log it
+	if !hasChartAnnotations(cluster) {
+		IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+		log.Info(fmt.Sprintf("Chart annotation not found for CAPI-based cluster. Cluster %s/%s will be ignored for deletion", cluster.Namespace, cluster.Name))
+		return nil
+	}
+
+	app := &gsapplication.App{}
+	if err := client.Get(ctx, getClusterAppNamespacedName(cluster), app); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, fmt.Sprintf("unable to get app CR for cluster %s/%s", cluster.Namespace, cluster.Name))
+		ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+		return nil
+	}
+	// ignore GitOps-managed resources, ensure we're not deleting cluster app CR of MC itself
+	if _, ok := app.Labels[fluxLabel]; ok {
+		IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+		log.Info(fmt.Sprintf("Found label %s in App CR. Cluster %s/%s will be ignored for deletion", fluxLabel, cluster.Namespace, cluster.Name))
+		return nil
+	}
+
+	log.Info(fmt.Sprintf("Cluster %s/%s has exceeded the default time to live (%s) and will be deleted", cluster.Namespace, cluster.Name, defaultTTL))
+
+	// delete App CR for the cluster
```

Comment:
[optional,minor]

We could skip ahead here if it already has a non-zero deletion timestamp

Reply:
I think we do that at top of reconcile:
https://github.com/giantswarm/cluster-cleaner/blob/35e45e2d18d72d49cd8f2d61464a57de486212bd/controllers/cluster_controller.go#L71-L76